### PR TITLE
CVSL-1555-Deactivate-in-progress-licences-after-CRD

### PR DESCRIPTION
**Pull request by jaleen**
Added a new job endpoint to inactivate the licences passed CRD. It retrieves licences and make the licences inactive

Also added  licence_event and audit_event auditing.****

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/repository/LicenceRepository.kt [hunk 1]
```diff
@@ -92,6 +92,19 @@ interface LicenceRepository : JpaRepository<Licence, Long>, JpaSpecificationExec
...
+
+  @Query(
+    """
+    SELECT l
+        FROM Licence l
+        WHERE l.statusCode  IN (
+            uk.gov.justice.digital.hmpps.createandvaryalicenceapi.util.LicenceStatus.IN_PROGRESS,
+            uk.gov.justice.digital.hmpps.createandvaryalicenceapi.util.LicenceStatus.SUBMITTED
+        )
+        AND COALESCE(l.conditionalReleaseDate, l.actualReleaseDate) < CURRENT_DATE
```
Comment (SebNegahbanMoJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/repository/LicenceRepository.kt:104]: I think the params for `COALESCE` should be reversed. ie:
`COALESCE( l.actualReleaseDate, l.conditionalReleaseDate) < CURRENT_DATE`
as we prioritise actualReleaseDate when determining the release date, and it's more likely to be null.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/repository/LicenceRepository.kt [hunk 2]
```diff
@@ -92,6 +92,19 @@ interface LicenceRepository : JpaRepository<Licence, Long>, JpaSpecificationExec
...
+    """,
+  )
+  fun getDraftLicencesPassedCRD(): List<Licence>
```
Comment (SebNegahbanMoJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/repository/LicenceRepository.kt:107]: Slight misnomer, perhaps `getDraftLicencesPassedReleaseDate` would be more accurate? Not a big deal though.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/jobs/DeactivateLicencesService.kt [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  fun deactivateLicencesJob() {
+    log.info("Job deactivateLicencesJob started")
+    val licencesToDeactivate = licenceRepository.getDraftLicencesPassedCRD()
+    if (licencesToDeactivate.isEmpty()) {
+      log.info("Job deactivateLicencesJob has no licences to time out")
+      return
+    }
+    log.info("deactivateLicencesJob is updating status INACTIVE on ${licencesToDeactivate.size} licences")
+    updateLicencesStatus(licencesToDeactivate)
+    log.info("TimeOutLicencesServiceJob updated status TIMED_OUT on ${licencesToDeactivate.size} licences")
```
Comment (SebNegahbanMoJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/jobs/DeactivateLicencesService.kt:39]: I think this log has been left in by mistake

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/jobs/DeactivateLicencesService.kt [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    licenceRepository.saveAllAndFlush(deactivateLicences)
+
+    deactivateLicences.map { licence ->
+      auditEventRepository.saveAndFlush(
+        AuditEvent(
+          licenceId = licence.id,
+          username = userName,
+          fullName = userName,
+          eventType = AuditEventType.SYSTEM_EVENT,
+          summary = "${reason ?: "Licence deactivated automatically as it passed CRD"} for ${licence.forename} ${licence.surname}",
```
Comment (SebNegahbanMoJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/jobs/DeactivateLicencesService.kt:59]: Similar to previous comment, `CRD` should probably be `release date` but not a big issue.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/jobs/DeactivateLicencesService.kt [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+      )
+
+      licenceEventRepository.saveAndFlush(
+        LicenceEvent(
+          licenceId = licence.id,
+          eventType = LicenceEventType.INACTIVE,
+          username = userName,
+          forenames = userName,
+          surname = userName,
+          eventDescription = "${reason ?: "Licence deactivated automatically as it passed CRD"} for ${licence.forename} ${licence.surname}",
```
Comment (SebNegahbanMoJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/jobs/DeactivateLicencesService.kt:71]: Similar to previous comment, `CRD` should probably be `release date` but not a big issue.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/integration/DeactivateLicencesTest.kt [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    val deactivatedLicences = webTestClient.post()
+      .uri("/licence/match")
+      .bodyValue(MatchLicencesRequest(status = listOf(LicenceStatus.INACTIVE)))
+      .accept(MediaType.APPLICATION_JSON)
+      .headers(setAuthorisation(roles = listOf("ROLE_CVL_ADMIN")))
+      .exchange()
+      .expectBodyList(LicenceSummary::class.java)
+      .returnResult().responseBody
+
+    assertThat(deactivatedLicences?.size).isEqualTo(1)
```
Comment (SebNegahbanMoJ) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/integration/DeactivateLicencesTest.kt:55]: May be worth adding a test case for `actualReleaseDate` being in the past to

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/repository/LicenceRepository.kt (hunk 1 above)
Comment (andrewrlee) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/repository/LicenceRepository.kt:104]: I wonder if we have the same issue on the time out job 👁️

Comment (andrewrlee) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/repository/LicenceRepository.kt:104]: looks ok on there: 
`    WHERE COALESCE(l.actualReleaseDate, l.conditionalReleaseDate) <= :timeOutDate`

Comment (jaleen) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/repository/LicenceRepository.kt]: removed coalesce as it wasn't useful when release date was not null and in future (theoretical possibility) and crd was in past

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/jobs/DeactivateLicencesService.kt (hunk 5 above)
Comment (jaleen) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/jobs/DeactivateLicencesService.kt]: changed to release date

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/jobs/DeactivateLicencesService.kt (hunk 3 above)
Comment (jaleen) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/jobs/DeactivateLicencesService.kt]: fixed

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/jobs/DeactivateLicencesService.kt [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  companion object {
+    private val log = LoggerFactory.getLogger(this::class.java)
+  }
+
+  @Transactional
+  fun deactivateLicencesJob() {
+    log.info("Job deactivateLicencesJob started")
+    val licencesToDeactivate = licenceRepository.getDraftLicencesPassedCRD()
+    if (licencesToDeactivate.isEmpty()) {
+      log.info("Job deactivateLicencesJob has no licences to time out")
```
[the remark this answers is not in the document]
Comment (jaleen) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/jobs/DeactivateLicencesService.kt]: thanks fixed

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/repository/LicenceRepository.kt (hunk 2 above)
Comment (jaleen) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/repository/LicenceRepository.kt]: updated to release date

**Review by andrewrlee — APPROVED** [review]
LGTM

**Pull request merged**